grace-core/src/Grace/Interpret.hs

```diff
@@ -72,31 +69,54 @@ interpretWith
     -> Input
     -> m (Type Location, Value)
 interpretWith bindings maybeAnnotation input = do
-    code <- case input of
-        Path file -> liftIO (Text.IO.readFile file)
-        Code text -> return text
+    let directory = case input of
+            Path file -> FilePath.takeDirectory file
+            Code _ _  -> "."
 
-    let name = case input of
-            Path file -> file
-            Code _    -> "(input)"
+    eitherExpression <- Except.runExceptT (Parser.parseInput input)
 
-    expression <- case Parser.parse name code of
-        Left message -> do
-            Except.throwError (ParseError message)
+    expression <- case eitherExpression of
+        Left e -> throwError (ParseError e)
+        Right expression -> return expression
 
-        Right expression -> do
-            let locate offset = Location{..}
+    interpretExprWith bindings maybeAnnotation directory expression
 
-            return (first locate expression)
+{- | Like `interpretWith`, but takes a pre-parsed expression instead of some
+     source code
+-}
+interpretExprWith
+    :: (MonadError InterpretError m, MonadIO m)
+    => [(Text, Type Location, Value)]
+    -- ^ @(name, type, value)@ for each custom binding
+    -> Maybe (Type Location)
+    -- ^ Optional expected type for the input
+    -> FilePath
+    -- ^ The base directory used when resolving filepath imports of relative paths
+    -> Syntax Location Import
+    -> m (Type Location, Value)
+interpretExprWith bindings maybeAnnotation directory expression = do
+    resolvedExpression <- for (annotate expression)
+        \(maybeAnnotation', input) -> case input of
+            Import.File file -> interpretWith bindings maybeAnnotation' (Path path)
+              where
+                path = FilePath.normalise (directory </> file)
+
+            Import.URI uri -> do
+                eitherResult <- liftIO do
+                    tryAny (Import.resolverToCallback Resolver.defaultResolver uri)
+
+                (importExpression, directory') <- case eitherResult of
+                    Left e -> throwError (ImportError uri (displayException e))
+                    Right (expr, Nothing) -> return (expr, directory)
+                    Right (expr, Just directory') -> return (expr, directory')
+
+                let relocate location = location
```

Comment:
I think you can avoid having to `relocate` the `Location` information if you change the `Input` type to add another constructor for `URI`.  In other words, if the `Input` that you pass to the `parseInput` function can have a `URI` constructor, then the `parseInput` function can ensure that the generated `Location` is correct and doesn't have to be amended.

As a bonus, if you do things this way then I believe you no longer need to split the `interpretWith` function into separate `interpretWith` / `interpretExprWith` functions because you can concentrate all of the fetching logic in one place.

Reply:
I am afraid I don't get the idea :slightly_frowning_face: Are you suggesting that the resolving takes place in `parseInput`? Or how should that function be aware of the actual text it is supposed to parse?

I am also not sure if we could (or even want to) get rid of the `interpretExprWith` function: It is used by `interpretWith` when parsing the main input fed to the interpreter but it is also interpreting the result of the resolvers/import callback.
And I'd like to keep the `Syntax Location Import` of those since I don't want to assume that the Grace expression which is imported is stored as a source code. It might as come from a DB or from a serialized CBOR file.

Reply:
I put up a pull request with my suggested change:

https://github.com/mmhat/grace/pull/1

The summary is that if you consolidate the `Import` and `Input` types into a single type then it simplifies several other things

Reply:
Ah, ok, that's cool! Thank you!